rust/main/chains/hyperlane-aleo/src/mailbox.rs

```diff
@@ -0,0 +1,240 @@
+};
+
+/// Aleo Ism
+#[derive(Debug, Clone)]
+pub struct AleoMailbox {
+    provider: AleoProvider,
+    address: H256,
+    program: String,
+    domain: HyperlaneDomain,
+}
+
+impl AleoMailbox {
+    /// Returns a new Mailbox
+    pub fn new(provider: AleoProvider, locator: &ContractLocator, conf: &ConnectionConf) -> Self {
+        Self {
+            provider,
+            address: locator.address,
+            program: conf.mailbox_program.clone(),
+            domain: locator.domain.clone(),
+        }
+    }
+
+    /// Get the ProgramID for a recipient address
+    /// `recipient` - H256 address of the recipient
+    ///
+    /// Returns ProgramID<CurrentNetwork>
+    ///
+    /// In aleo there is no direct way to map from an H256 address to a ProgramID.
+    /// Instead we look up the address in the `registered_applications` mapping of the mailbox program.
+    /// This mapping stores the ProgramID for each registered recipient address.
+    async fn get_recipient(&self, recipient: H256) -> ChainResult<ProgramID<CurrentNetwork>> {
+        let aleo_address = Address::<CurrentNetwork>::from_bytes_le(recipient.as_bytes())
+            .map_err(HyperlaneAleoError::from)?;
+        let program_id: [u8; 128] = self
+            .provider
+            .get_mapping_value(&self.program, "registered_applications", &aleo_address)
+            .await
+            .map_err(|_| {
+                HyperlaneAleoError::Other(format!(
+                    "Expected recipient to be registered, but was not: {aleo_address}",
+                ))
+            })?;
+        let program_id =
+            CStr::from_bytes_until_nul(&program_id).map_err(HyperlaneAleoError::from)?;
+        let program_id = program_id.to_str().map_err(HyperlaneAleoError::from)?;
+        Ok(ProgramID::from_str(program_id).map_err(HyperlaneAleoError::from)?)
+    }
+
+    /// Get the arguments for processing a message
+    /// `message` - HyperlaneMessage to process
+    /// `metadata` - Metadata bytes needed to process the message
+    /// Returns Vec<String> of arguments
+    ///
+    /// The metadata is padded or truncated to 512 bytes as required by the Aleo contracts.
+    /// The message ID is converted to little endian u128 as required by the Aleo contracts.
+    /// The ISM is looked up from the recipient's app metadata.
+    /// The app metadata is also looked up from the recipient address.
+    async fn get_process_args(
+        &self,
+        message: &HyperlaneMessage,
+        metadata: &[u8],
+    ) -> ChainResult<Vec<String>> {
+        let message_length = message.to_vec().len() as u32;
+        let aleo_message: AleoMessage = message.clone().into();
+
+        // Pad or truncate metadata to 512 bytes
+        let copy_len = metadata.len().min(512);
+        // Initialize with 1s as Aleo will error on zero bytes for the ECDSA recovery
+        let mut buf = [1u8; 512];
+        buf[..copy_len].copy_from_slice(&metadata[..copy_len]);
+        let metadata = buf;
```

Comment:
Would it be worth to extract padding into a separate function in `utils` and add tests?